### PR TITLE
Conversion to Typescript. Closes #150

### LINKED ISSUE #150: Typings for Typescript
I'm going to provide typings for Typescript. There are two common ways of doing it:
1. Add `d.ts` file(s) to the library itself (the way `mobx` does, for example).
2. Publish typings under [DefinitelyTypes](http://definitelytyped.org/) a.k.a. [@types](https://www.npmjs.com/~types).

I'd prefer the first one. @foxhound87, any objections?

### PR DESCRIPTION
**Pull request by wwahammy**
As mentioned in https://github.com/foxhound87/mobx-react-form/pull/362#issuecomment-411487721, I've converted Mobx-react-form to Typescript. This was a mostly straightforward conversion from es6 to Typescript. Important things to note:

* All the tests pass on Mobx3-5
* Typings are generated and included
* tslint replaces eslint. We use the typescript version of the AirBnb style guide.
  * Limitation: tslint doesn't currently check the test directory by default so some of the tests don't pass. It's something to be added in the near future but I wanted to submit the PR :)
* I tried to minimize the formatting changes so the diffs weren't a mess going forward. There was some judgment involved so it was just my best effort on that. Nonetheless, I'm happy with the result.
* I didn't see an obvious way to get the coverage tool to work with typescript. It still runs as part of the build but gives a useless result.
* Almost every complex type is set to `any`. The type system is somewhat complex in mobx-react-form so I figured it was best to just get this done this way and let us improve over time.

I'm adding a few line comments in the pull request to explain some decisions.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -68,37 +69,36 @@
...
     "cross-env": "^5.1.4",
     "cz-conventional-changelog": "^2.0.0",
-    "eslint": "4.11.0",
-    "eslint-config-airbnb-base": "12.1.0",
-    "eslint-plugin-import": "2.8.0",
     "husky": "0.13.1",
     "json-loader": "0.5.4",
-    "mobx5": "npm:mobx@^5.0.0",
-    "mobx4": "npm:mobx@^4.0.0",
+    "mobx": "npm:mobx@^4.0.0",
```
Comment (wwahammy) [inline package.json:86]: Mobx provides its own types so we need to have a way to get the types when we do a build. We do that here.

File: src/Base.ts [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+  $isEvent,
+  hasIntKeys } from './utils';
+import Actions from './shared/Actions';
+import Events from './shared/Events';
+import Helpers from './shared/Helpers';
+import Initializer from './shared/Initializer';
+import Utils from './shared/Utils';
+
+export default abstract class Base implements Actions, Events, Helpers, Initializer, Utils {
+  abstract invalidate(message?: any, async?: boolean):any;
```
Comment (wwahammy) [inline src/Base.ts:15]: If we don't put these in as abstract, nothing compiles. It's kind of messy but does work.

File: src/Field.ts [hunk 3]
```diff
@@ -12,6 +19,8 @@ import {
   parseInput,
   defaultClearValue } from './parser';
 
+const mobxAny = mobx as any;
```
Comment (wwahammy) [inline src/Field.ts:22]: To get the asMap function from Mobx3 we need to have an effectively untyped version of Mobx. We do that here.

File: src/Field.ts [hunk 4]
```diff
@@ -43,7 +52,10 @@ const setupDefaultProp = (instance, data, props, update, {
 
 export default class Field extends Base {
 
-  fields = observable.map ? observable.map({}) : asMap({});
+  fields:ObservableMap<any> = observable.map ?
+    observable.map({}) :
+    mobxAny.asMap({});
```
Comment (wwahammy) [inline src/Field.ts:57]: Here's where we use the untyped Mobx

File: src/shared/Actions.ts [hunk 5]
```diff
@@ -1,17 +1,30 @@
...
+import * as _ from 'lodash';
+import * as utils from '../utils';
+import * as parser from '../parser';
+import { CommonProperties } from './CommonProps';
 
 /**
   Field Actions
 */
-export default {
+class SharedActions implements CommonProperties {
```
Comment (wwahammy) [inline src/shared/Actions.ts:10]: All of the shared files have to include the properties and functions they use from other shared files. I didn't want to throw everything together into one file so I include the necessary functions and properties in CommonProperties and then pull that into each shared file. It's not the prettiest but it gets the job done.

**Comment by wwahammy** [comment]
While I doubt it was a major concern to anyone but me, this version seems to work on IE9-10 with polyfills. For reasons I couldn't figure out, the previous one didn't. *shrug*

Again, I doubt a major issue for anyone else but it's a nice perk :)

**Comment by foxhound87** [comment]
Great @ericschultz , I merged your PR in the `typescript` branch.
https://github.com/foxhound87/mobx-react-form/tree/typescript

**Comment by wwahammy** [comment]
Thanks a ton @foxhound87! What kind of schedule do you anticipate for migrating this to master? I just want to know how much work it's going to be to keep up :)

**Comment by foxhound87** [comment]
Working directly on the typescript code for me would be slower, and for now I prefer stick to ES6 to provide the updates and fixes. Are you willing to keep the `typescript` branch in sync with my future commits? Then I think I will merge it to master when the major tasks will be done.

**Comment by foxhound87** [comment]
See the tasks here: https://gist.github.com/foxhound87/f176f19054148647e2a38d60cc09259b

Mainly I want first implement:
- Splitting Validators in multiple Validation Drivers packages
- Plugins & Drivers management (enable connection of Validation Drivers to Validation Plugins).

**Comment by wwahammy** [comment]
@foxhound87 I'll do my best. It's not a ton of work but just something to commit to regularly. I just want to avoid this getting abandoned since it has some pretty big benefits for people using Typescript. It also caught a few bugs!

**Comment by jonesnc** [comment]
@ericschultz @foxhound87 Any updates on this? I'd love to start using this.

**Comment by foxhound87** [comment]
There are new updates to be committed.

File: src/shared/Actions.ts [hunk 6]
```diff
@@ -20,7 +33,7 @@ export default {
   submit(o = {}) {
     this.$submitting = true;
 
-    const exec = isValid => isValid
+    const exec:any = isValid => isValid
```
Comment (KODerFunk) [inline src/shared/Actions.ts:36]: No `any` please! There `(isValid: boolean) => Function | Promise`,
and yes bug at 60 at `return utils.isPromise(exec) ? ... : ...`;
And!
My promise must call after success validation! No before!

Comment (foxhound87) [reply src/shared/Actions.ts:36]: Can you clarify how it should be in the code?

**Pull request closed without merging**